Themes: Update Preview and Try & Customize workflow to improve UX

abotic opened the pull request:
# Update Preview and Try & Customize theme workflow
Related to:
- https://github.com/Automattic/wp-calypso/issues/92197

## Proposed Changes
- Modified theme preview behavior to show preview modal instead of direct redirect
- Updated redirect behavior for "Try & Customize" to use site-editor URL format consistently
- Unified preview button rendering with consistent design and translations
- Improved code organization for theme preview related functionality

## Why are these changes being made?
Currently, the theme preview workflow is inconsistent between different theme types and the "Try & Customize" button redirects to different URLs based on theme type. This update provides a more consistent experience by:
1. Always showing the preview in a modal first
2. Using the site-editor URL format for "Try & Customize" redirects
3. Maintaining consistent button labeling and behavior

## Testing Instructions
1. Log into your WordPress.com account
2. Navigate to Themes section
3. Test the following scenarios:

| Scenario | Steps | Expected Behavior |
|----------|-------|------------------|
| Preview button | Click "Preview" on any theme (except active one) | Opens preview modal with "Try & Customize" option |
| Try & Customize | Click "Try & Customize" in preview modal | Redirects to site-editor.php with correct theme preview URL |
| Button consistency | View different theme types | Preview button appears consistently with same styling |
| Logged out | View themes while logged out | Preview works and prompts for login when needed |
| No site context | View themes without selecting a site | Preview works and prompts for site selection |
| Active theme | View currently active theme | Shows appropriate preview and customize options |

| Before | After |
| ---- | ---- |
| <video src="https://github.com/user-attachments/assets/5aabbf1a-5fd1-4ad7-91ac-820bae707fba" controls width="250"></video> | <video src="https://github.com/user-attachments/assets/e503642b-7e9a-45c2-82a5-cd11d41e4413" controls width="250"></video> |


## Pre-merge Checklist
- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] Have you written new tests for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

abotic committed:
Themes: Update Preview and Try & Customize workflow to improve UX

abotic committed:
Fix try and customize for bundled/woocommerce themes

mmtr reviewed (commented):
Nice work @abotic and great testing instructions!

I noted a small regression when testing on an Atomic site:

- Go to `/themes` while logged in
- Pick a theme
- Select an Atomic site
- Click on the Preview button
- Select "Try and customize"
- :x: The modal displays the theme slug (e.g. pub/leancv) instead of the theme name

Before | After
--- | ---
<img width="1272" alt="Screenshot 2025-01-14 at 14 27 35" src="https://github.com/user-attachments/assets/3195a141-a82d-4cf5-9a3e-f2a85042d5a9" /> | <img width="1275" alt="Screenshot 2025-01-14 at 14 27 48" src="https://github.com/user-attachments/assets/16419b53-2b57-410c-9c16-bf4d520f9036" />

abotic committed:
handle atomic site navigation

abotic commented:
> Nice work @abotic and great testing instructions!
> 
> I noted a small regression when testing on an Atomic site:
> 
> * Go to `/themes` while logged in
> * Pick a theme
> * Select an Atomic site
> * Click on the Preview button
> * Select "Try and customize"
> * ❌ The modal displays the theme slug (e.g. pub/leancv) instead of the theme name
> 
> Before	After
> <img alt="Screenshot 2025-01-14 at 14 27 35" width="1272" src="https://private-user-images.githubusercontent.com/1233880/402955553-3195a141-a82d-4cf5-9a3e-f2a85042d5a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc1NTM0NDAsIm5iZiI6MTczNzU1MzE0MCwicGF0aCI6Ii8xMjMzODgwLzQwMjk1NTU1My0zMTk1YTE0MS1hODJkLTRjZjUtOWEzZS1mMmE4NTA0MmQ1YTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjJUMTMzOTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWQ2NWRmYjY4MWExYWZkNDdhOTI5NDhjY2Q0NmU4MWQ0YWIxZDEyZTZkNTQwYTZmNThmOTFmNDgxNjlhNmM3MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.tQxyjTnADr6rypyXnDXYTg5MxLKt3IZowVHpIA6F60I">	<img alt="Screenshot 2025-01-14 at 14 27 48" width="1275" src="https://private-user-images.githubusercontent.com/1233880/402955528-16419b53-2b57-410c-9c16-bf4d520f9036.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc1NTM0NDAsIm5iZiI6MTczNzU1MzE0MCwicGF0aCI6Ii8xMjMzODgwLzQwMjk1NTUyOC0xNjQxOWI1My0yYjU3LTQxMGMtOWMxNi1iZjRkNTIwZjkwMzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjJUMTMzOTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDU5ZDJhNWY4YTFhNTAyMmM1NjkwNTU2YzJiMWVjNjBlZjVhNzIyNjMwNjEyNmVmYzI4MzkyMjgzMTM4YjAwMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.PtNFuOS-4c0R3dM-lwHKycU6rFexX2_XXXVrb4DoM7M">

Nice catch @mmtr !
I've fixed this in my last commit.
Let me know if you find any other issue.

mmtr commented on client/components/web-preview/toolbar.jsx:
```diff
@@ -78,7 +83,33 @@ class PreviewToolbar extends Component {
+
+			const isBundledTheme = themeType === BUNDLED_THEME;
+			const themePrefix = isBundledTheme ? 'premium' : 'pub';
+			const themeFilter = isBundledTheme ? '?tab_filter=recommended&tier_filter=woocommerce' : '';
```
Why is the tab filter hardcoded to `recommended` and the tier filter to `woocommerce`? Shouldn't we use the current tab and tier?

Also, not every bundled theme belongs to the WooCommerce bundle. We also have the Sensei LMS bundle which includes the Course theme, and this theme is not part of the WooCommerce tier.

See the screen recording below and note how I start on:
- Category: All
- Tier: All

And after selecting Course and going back, I land on:
- Category: Recommended
- Tier: WooCommerce


https://github.com/user-attachments/assets/9b4b391c-87fa-47e7-a956-43fd6ebb3da6

abotic replied:
Makes complete sense, thank you for looking at it. I've changed the approach and we should not rely on hardcoded values.

mmtr commented on client/components/web-preview/toolbar.jsx:
```diff
@@ -78,7 +83,33 @@ class PreviewToolbar extends Component {
+			}
+
+			const isBundledTheme = themeType === BUNDLED_THEME;
+			const themePrefix = isBundledTheme ? 'premium' : 'pub';
```
We have some `premium/` themes which are not bundled. Example: `premium/tenku` – https://wordpress.com/theme/tenku

<img width="1392" alt="Screenshot 2025-01-22 at 19 09 34" src="https://github.com/user-attachments/assets/e14734e8-3b6f-4499-b8da-8503239cff8f" />

This causes the "Try & Customize" to show a wrong theme:

<img width="1274" alt="Screenshot 2025-01-22 at 19 11 23" src="https://github.com/user-attachments/assets/7289424e-ddf0-4d95-9403-632b0c88837f" />

abotic replied:
Was not aware of that combination, I've used stylesheet property from now on so we don't need to construct theme prefix on our own. Thanks

mmtr reviewed (commented):
Something is still not working on my Atomic site.

If I try & customize LeanCV, I get my current theme (Course) instead:

https://github.com/user-attachments/assets/d09c2d5d-65a4-4e14-930c-7e527ac7eeda

abotic committed:
fix origin for atomic sites, use proper theme prefix

abotic committed:
for atomic sites, install theme if needed before navigating to site editor

abotic committed:
resolve conflicts

abotic commented:
> Something is still not working on my Atomic site.
> 
> If I try & customize LeanCV, I get my current theme (Course) instead:
> 
>  Screen.Recording.2025-01-22.at.19.14.16.mov

@mmtr  Thanks for reviewing this PR and pointing out a behavior that I couldn't reproduce locally. The main issue (which we discussed on Slack) was how we should handle atomic sites navigation to the site editor.

I couldn't reproduce some of these issues locally (and initially pushed a non-working version) because I had already (unknowingly) installed themes on the atomic site I used on production and localhost, and did most of my testing on new and existing simple sites.

On atomic sites, when we click “Try and customize” for a theme that isn’t already installed, we first need to install it before continuing with the redirect. Simple sites don’t require that step, as themes come pre-installed. This was the root cause of most of the problems, but everything should now work as expected according to the instructions and video. Thank you for testing this!

xavier-lc commented on client/my-sites/theme/main.jsx:
```diff
@@ -1032,6 +1008,36 @@ class ThemeSheet extends Component {
+			return null;
+		}
+
+		const buttonProps = {
```
Out of curiosity, why use this `buttonProps` object and then spread it instead of just assigning the props to `Button`as usual?

abotic replied:
No particular reason. I did it more for the readability of all those ternary cases. But it's better to stick to the styling we use in other places, so I'll just assign the props to the Button. Thanks!

xavier-lc commented on client/my-sites/theme/main.jsx:
```diff
@@ -1032,6 +1008,36 @@ class ThemeSheet extends Component {
+	renderPreviewButton = () => {
+		const { isLivePreviewSupported, translate, isExternalLink } = this.props;
+
+		if ( ! isLivePreviewSupported && ! this.shouldRenderPreviewButton() ) {
+			return null;
+		}
```
Before the changes, the render condition was `this.shouldRenderPreviewButton() && ! isLivePreviewSupported && <Button ... />`, so I'm wondering if this code should be:

```
		if ( isLivePreviewSupported || ! this.shouldRenderPreviewButton() ) {
			return null;
		}
```

abotic replied:
Nice catch! Thanks

abotic replied:
Actually, we won't need `isLivePreviewSupported` as we will always show the preview modal now. So removed that entirely.

xavier-lc replied:
Cool!

mmtr commented on client/my-sites/theme/main.jsx:
```diff
@@ -1032,6 +1008,36 @@ class ThemeSheet extends Component {
 	};
 
+	renderPreviewButton = () => {
+		const { isLivePreviewSupported, translate, isExternalLink } = this.props;
```
Where is `isExternalLink` defined now?  

I see you removed `const isExternalLink = ! this.props.isWpcomTheme || this.props.isExternallyManagedTheme;`, but there is no replacement for that 🤔

abotic replied:
Good point, haven't noticed any difference in "Preview" button while testing, but now looking at regression - we actually need that otherwise we won't have that external icon .

mmtr reviewed (commented):
Pretty close! Overall this works quite well, but there are some small regressions when compared to the current `trunk`

### Issue 1: Navigation back is broken.

**Before**

✅ When going back I land in the theme details page.
✅ If click on the "Themes" menu in the breadcrumb menu, I land in the theme showcase and the selected tier is still the same I had before

https://github.com/user-attachments/assets/a7eed5f0-b020-4c5f-9e93-ca6967789de0

**Now**

:x: When getting back, I land in the theme showcase.
:x: The selected tier is not the same I had before.

https://github.com/user-attachments/assets/4a0b29cc-3427-4cef-b35d-1fc301f05f22

### Issue 2: No loading state

**Before**

✅ When previewing a theme, the "Preview & Customize" button changed to a loading state to indicate that there is something in progress:

https://github.com/user-attachments/assets/58f36417-784e-4b70-b4de-b00617f53d85

**Now**

:x: When previewing a theme, nothing happens after clicking on the "Try and customize" during a few seconds. It's not obvious that users have to wait a little bit, and the page eventually will load the preview.


https://github.com/user-attachments/assets/70104286-2877-4b55-b97a-cf3643b82323

### Issue 3: "Vibrating" preview

**Before**

✅ When loading the details on the Tenku theme, the preview is displayed normally on the right

https://github.com/user-attachments/assets/455c8252-5f56-44a8-8e2c-02656516eb4f

**Now**

❌  When loading the details on the Tenku theme, the preview "vibrates". Not sure if this is an issue related to your changes, but I cannot reproduce it on `trunk`, only in this PR. It might be also specific to my screen size (1280px width).


https://github.com/user-attachments/assets/07625236-1826-498d-a676-c08eae3e26bb



### Issue 4: "Try and customize" visible in site preview

**Before**

✅ When previewing my site in `/view/:site`, the header has the expected elements:

<img width="1276" alt="Screenshot 2025-02-10 at 16 39 07" src="https://github.com/user-attachments/assets/0f168050-41c4-413c-8475-85e3f2a84cc8" />

**Now**

❌ When previewing my site in `/view/:site`, the header has an unexpected "Try and customize" button:

<img width="1268" alt="Screenshot 2025-02-10 at 16 39 13" src="https://github.com/user-attachments/assets/11b4f013-4cfc-4677-8ce4-2627634e6ec5" />

abotic committed:
Fixes comments, navigations, loading state, other places view etc

abotic committed:
resolve conflicts

abotic committed:
Remove the usage of isLivePreviewSupported as we will always show it

abotic commented:
@mmtr Thanks for such a great regression testing. Everything has been fixed except "Vibrating Tenku theme". I've tried with several screen sizes and wasn't able to reproduce this. Maybe I've fixed it along the way as I've changed some CSS to match the trunk in some minor tweaks.

Additionally, about the loading state: 
`Try and customize` is borderless and is shown as link, so we can't really show loading state on the "button" itself nicely.
I've added a spinner after the button has been clicked, do you think this is good or do you have some other idea?

mmtr commented on client/components/web-preview/component.jsx:
```diff
@@ -69,6 +69,7 @@ export class WebPreviewModal extends Component {
 		disableTabbing: false,
 		showExternal: true,
 		showClose: true,
+		showEditHeaderLink: true,
```
Will this cause any unexpected side effect to all the code currently using the `WebPreview` component? 

https://github.com/search?q=repo%3AAutomattic%2Fwp-calypso%20%22%3CWebPreview%22&type=code

If this is only needed for the `ThemePreview` component, maybe it's safer to explicitly override the `showEditHeaderLink` prop [here](https://github.com/Automattic/wp-calypso/blob/1d62e7cf54dcd97add271a5b8017f742c7b980c6/client/my-sites/themes/theme-preview.jsx#L73)?

abotic replied:
While doing regression testing, it didn't cause any issues on other places, but I completely agree that we should simply override it to prevent any issues on other places. Thanks!

mmtr commented on client/components/web-preview/toolbar.jsx:
```diff
@@ -68,17 +80,30 @@ class PreviewToolbar extends Component {
+
+		const { isAtomic, selectedSiteId, siteEditorUrl, themeInstallId } = this.props;
+
+		// For atomic sites, we need to install theme before navigating to site editor
+		// If theme is already installed, installation will silently fail, and we just switch to the site-editor.
+		try {
+			if ( isAtomic ) {
+				await this.props.installTheme( themeInstallId, selectedSiteId );
+			}
+			window.location.href = siteEditorUrl;
+		} catch ( error ) {
+			this.setState( { isRedirecting: false } );
+		}
```
Nitpick (non-blocking): Maybe we can check whether the theme is installed and avoid an unnecessary install action if the theme is already installed?

abotic replied:
@mmtr I thought about this initially. Looking at the code, `installTheme` is already handling the case of installed themes correctly. When a theme is already installed, the request simply returns:

```
    "body": {
        "error": "theme_already_installed",
        "message": "The theme is already installed"
    }
```

and we proceed with the redirect to site-editor as expected.

Adding an additional check like `!!getTheme(state, siteId, ownProps.themeId)` would require checking this on all sites, even non-atomic ones, or we'd end up with nested conditionals before proceeding with `installTheme`. Since it's already handled properly to silently fail and redirect to the site editor, the extra check seems unnecessary.

The catch block here is specifically for handling actual issues that could arise during theme installation, such as network issues, server errors, or permission problems.

mmtr reviewed (commented):
Works wonderfully, thanks for addressing all the feedback!

The only issue I found is that classic themes have the "Try and customize" button, but they shouldn't because they cannot be previewed in the site editor (only block themes can be previewed).

To reproduce:
- Go to `/theme/twentyseventeen/:site`
- Preview the theme
- :x: The "Try and customize" is visible, but it shouldn't because Twenty Seventeen is a classic theme. If you click on it, you'll see that the theme is not loaded correctly in the site editor

Also left a couple of minor questions/suggestions

mmtr commented on client/components/web-preview/toolbar.jsx:
```diff
@@ -158,12 +183,20 @@ class PreviewToolbar extends Component {
 					{ showEditHeaderLink && canUserEditThemeOptions && (
 						<Button
```
Related to my comment above, since this is the new "Live Preview" button, you should check that the "Live Preview" feature is available.

For instance, [classic themes don't support the "Live Preview" feature](https://github.com/Automattic/wp-calypso/blob/ccc502532b765b0025ce2b9302de08f1fa1742d4/client/state/themes/selectors/get-is-live-preview-supported.ts#L103-L107).

abotic replied:
Fixed, thanks.

mmtr commented:
> I've added a spinner after the button has been clicked, do you think this is good or do you have some other idea?

That looks great, thank you!

abotic committed:
handle live preview for classic themes and override edit header link

mmtr reviewed (approved):
Thanks for your hard work on this 🚀

BogdanUngureanu commented:
Non-blocking: I did a quick test and noticed that "Preview" and "Try and customize" links don't have tracks events.

@abotic You can add them in a follow-up PR so that it won't block deploying this PR.

abotic commented:
> Non-blocking: I did a quick test and noticed that "Preview" and "Try and customize" links don't have tracks events.
> 
> @abotic You can add them in a follow-up PR so that it won't block deploying this PR.

The tracking events are at place for both actions, @BogdanUngureanu and me double checked.
